Softeq.XToolkit.PushNotifications/PushTokenSynchronizer.cs

```diff
@@ -0,0 +1,161 @@
+        {
+            get => _internalSettings.GetValueOrDefault(_isTokenSavedOnServerKey, default(bool));
+            set => _internalSettings.AddOrUpdateValue(_isTokenSavedOnServerKey, value);
+        }
+
+        private bool IsTokenRegisteredInSystem
+        {
+            get => _internalSettings.GetValueOrDefault(_isTokenRegisteredInSystemKey, default(bool));
+            set => _internalSettings.AddOrUpdateValue(_isTokenRegisteredInSystemKey, value);
+        }
+
+        /// <inheritdoc />
+        public Task SynchronizeTokenIfNeededAsync()
+        {
+            if (!IsTokenRegisteredInSystem && IsTokenSavedOnServer)
+            {
+                return UnregisterFromRemotePushNotificationsAsync();
+            }
+
+            if (IsTokenRegisteredInSystem && !IsTokenSavedOnServer)
+            {
+                return DoSendTokenToServer(_pushTokenStorageService.PushToken);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        /// <inheritdoc />
+        public async Task UnregisterFromRemotePushNotificationsAsync()
+        {
+            IsTokenRegisteredInSystem = false;
+
+            if (!IsTokenSavedOnServer)
+            {
+                return;
+            }
+
+            var tokenRemovedFromServer = await _remotePushNotificationsService
+                .RemovePushNotificationsToken(_pushTokenStorageService.PushToken)
+                .ConfigureAwait(false);
+            if (tokenRemovedFromServer)
+            {
+                IsTokenSavedOnServer = false;
+            }
+        }
+
+        /// <inheritdoc />
+        public void OnTokenChanged(string token)
+        {
+            if (_pushTokenStorageService.PushToken == token)
+            {
+                return;
+            }
+
+            if (string.IsNullOrWhiteSpace(token))
+            {
+                OnRegisterFailedInternalAsync().FireAndForget(_logger);
+            }
+            else
+            {
+                OnRegisterSuccessInternalAsync(token).FireAndForget(_logger);
+            }
+        }
+
+        /// <inheritdoc />
+        public async Task OnRegisterFailedInternalAsync()
+        {
+            await UnregisterFromRemotePushNotificationsAsync()
+                .ConfigureAwait(false);
+
+            _pushNotificationsHandler.OnPushRegistrationCompleted(
+                false,
+                IsTokenSavedOnServer);
+        }
+
+        private async Task OnRegisterSuccessInternalAsync(string token)
+        {
+            IsTokenRegisteredInSystem = true;
+
+            if (!IsTokenSavedOnServer)
+            {
+                _pushTokenStorageService.PushToken = token;
+
+                await DoSendTokenToServer(token)
+                    .ConfigureAwait(false);
+            }
+
+            _pushNotificationsHandler.OnPushRegistrationCompleted(
+                true,
+                IsTokenSavedOnServer);
+        }
+
+        private async Task DoSendTokenToServer(string token)
```

Comment:
Why can't we move this retry logic to `IRemotePushNotificationsService` implementation?

Reply:
Retry logic can be implemented on both sides, I guess, but `IRemotePushNotificationsService` doesn't know about token lifecycle, so it won't be abele to stop retrying to send token if we already "unregistered" this token.
`IPushTokenSynchronizer` does know about token lifecycle, so it can stop retrying on "unregister" (although it is not implemented yet)